### PR TITLE
machines: Implement creation of Virtual Networks

### PR DESCRIPTION
**Pull request by skobyda**
https://bugzilla.redhat.com/show_bug.cgi?id=1672755

### CODE REVIEW

**Comment by skobyda** [comment]
**This PR is not finished yet and not ready for review**. I would just like some opinions about design from @andreasn or @garrett, to know if I'm going in right direction. Mostly on stuff shown in screenshots and described below.

So, basic properties which you can set for Virtual network are (apart from standart stuff as connection and name): Forward mode, physical device (only for specific forward modes) and IP configuration.
Screenshot from 2019-07-09 14-52-26
Screenshot from 2019-07-09 14-52-46

So the thing is whole IP configuration stuff is optional. But IP configuration section has another optional "subconfiguration": dhcp range. So it's basically nested optional configuration. And I don't know if I represent that correctly. Right now there is a checkbox, which if it's checked, will open a whole IP configuration section, and inside of that is another checkbox which will open whole DHCP configuration section.
Screenshot from 2019-07-09 14-53-23

**Comment by garrett** [comment]
Are IPv4 and IPv6 manual configuration? And if they're not enabled, it would be automatic (via DHCP, I would guess)? And would automatic be _typical_?

If so, perhaps we might consider something like an expandable area for setting up "Show Static Settings" (or something like this), and it would be similar to the expander in Virtual Machine's → VM → disks → add disk dialog:

Screenshot_2019-07-09 Virtual Machines - Sunny

Screenshot_2019-07-09 Virtual Machines - Sunny(1)

**Comment by garrett** [comment]
@andreasn and I talked about this during our call and we came up with an idea for a UI:


Label  | Dropdown options  |  Notes
-- | -- | --
  |   |  
Address assignment | DHCP | (default)
  | DHCP range |  
  | Static IPv4 |  
  | Static IPv6 |  
  | Static IPv4 + IPv6 |  

It would have a label (perhaps "Address assignment"?) and then a dropdown. After the dropdown, the appropriate options would be shown.

Do you think this would work? If not, why not?

**Comment by skobyda** [comment]
> Are IPv4 and IPv6 manual configuration? And if they're not enabled, it would be automatic (via DHCP, I would guess)? And would automatic be _typical_?

Yes those are manual configuration. Automatic configuration doesn't exist in libvirt virtual networks, since libvirt has no way how to automatically calculate it. Either you have manually configure IP address, or you use forward mode where IP configuration doesn't make sense (and is not allowed).
If you IP configuration is allowed, DHCP inside virtual network will assign IP addresses to VMs out of whole subnet's range or range you define in DHCP Range option.
Any stuff related to host's subnet and physical network is outside of this scope and should not be present here.

You see IP configuration depends on forward mode. In case of forward mode:
- **NAT and Open**, you  **have to define IP configuration** (either IPv4 or IPv6 or both)
- *Vepa and Bridge**, you **cannot define IP configuration** (it doesn't make sense and I think libvirt will not allow it)
- no forward mode (**Isolated network**), **both cases are available**. You can define IP configuration or you won't define any and then define static IPs for each domain.
So in summary, depending on forward modes, either there needs to be IP configuration, there cannot be any, or IP configuration is optional.

Then **DHCP range option.** **DHCP configuration is always part of IP configuration.**
- DHCP range is always available everywhere where you can define IP configuration.
- And it's always optional. If not configured, libvirt will assume whole subnet's range.
- And there is separate DHCP configuration for IPv4 and IPv6.

> @andreasn and I talked about this during our call and we came up with an idea for a UI:
> 
> Label 	Dropdown options 	 Notes
> Address assignment	DHCP	(default)
>  	DHCP range	 
>  	Static IPv4	 
>  	Static IPv6	 
>  	Static IPv4 + IPv6	 
> It would have a label (perhaps "Address assignment"?) and then a dropdown. After the dropdown, the appropriate options would be shown.
> 
> Do you think this would work? If not, why not?

Dropdown could work, but DHCP Range option kinda doesn't fit in there, and it misses "No IP configuration option". So I would rather imagine dropdown having four option: No IP configuration, IPv4, IPv6, IPv4 + IPv6. These option would however be dynamically changed depending what forward mode user uses. E.G. if he chooses NAT or Open, he looses "No IP configuration Option".
I don't think DHCP range should be represented in this dropdown, because DHCP range configuration heavily depends on IP configuration. DHCP configuration is not of equal status to IP configuration if you know what I mean.

**Comment by andreasn** [comment]
> No IP configuration, IPv4, IPv6, IPv4 + IPv6.

This sounds like a good idea to me.

**Comment by garrett** [comment]
> if he chooses [...] he loses [….]

You shouldn't presume someone is male, unless you're talking about a specific person who identifies as male. It should be changed to "he or she" (or "she or he")... or simply "they". Alternatively, "the user" (although it's problematic to say "user" for all sorts of reasons... so in this case, I usually say "the person using..." or "the person setting up the network" or "the administrator" or such), or "someone".

This matters the most in our UI, docs, and website... but it also matters a lot in conversations on issues and pull requests... and also on IRC.

It's important because:
- Not all of the people using our software are male.
- Not everyone on the team is male.
- We want to be inclusive of everyone.

Just because you and I are male, doesn't mean we should presume others are by default. In fact, it's even more important we should be sensitive and watch our word choices because the industry (and society in general even) is biased torward males.

I understand different languages treat pronouns differently, so it could be a slip-up when writing in English, but it's really important we get this right.

> No IP configuration

What happens when there's no IP configuration? Do things work without manually entering IPs? What's the best way to have defaults that lets someone successfully complete this task?

**Comment by skobyda** [comment]
> I usually say "the person using..." or "the person setting up the network" or "the administrator" or such), or "someone".

Good idea, I also often just say "user". This occassion was just slip of a tongue caused by grammar of my native language.

> What happens when there's no IP configuration? Do things work without manually entering IPs? 

It heavily depends on forward modes. Some need IP configuration, while others have no use for IP configuration so they forbid it.

**Vepa** virtual network: uses macvtaps, which directly connects physical interface with VM's interface, so there is no need to set IP of virtual network. If you do pass IP configuration, libvirt will return an error, so I think we shouldn't support it in the UI either for this mode.
**Bridge** virtual network: uses host's bridge, to connect VMs to physical interface. I believe the reason why you don't need to use IP configuration is because Bridge works on Data Link Layer of OSI model, where IPs do not exist(?). So libvirt wouldn't use IP configuration here, too. But I'm not sure if that's the reason libvirt doesn't support IP configuration for this forward mode. If you do pass IP configuration, libvirt will return an error, so I think we shouldn't support it in the UI either for this mode.
**NAT and Open** virtual networks: they require IP configuration. If not inputted, libvirt will return error, so we should enforce it in the UI too for these modes.

> What's the best way to have defaults that lets someone successfully complete this task?

For **forward modes**: would say NAT is most used one and should be default one. Althought that's just my opinion.

For **IP configuration dropdown**: For forward modes which allow IP configuration, default option should be IPv4. If forward mode doesn't allow IP configuration, whole dropdown and IP configuration should be hidden.

For **IP configuration values**: virt-manager defautly puts address and netmask/prefix, so I can look how do they calculate it and do it similar way.
**DHCP** shouldn't be defaultly disabled, as it's additional option.

So in summary, defaults for basic Network creation should be: 
forward mode: NAT, IP configuration: IPv4 with address and prefix already filled out, and with DHCP disabled. This will create I would say very common basic virtual network, with user needing to input just network's name. But these default are just my opinion and perhaps others could disagree.

**Comment by andreasn** [comment]
The hierarchy of elements, as what is a child of what was hard for me to read, so I tried to address it like this:
mockup

**Comment by garrett** [comment]
@andreasn: It looks nice. However, my big issue is with the placeholder text. Placeholders should not be used as labels and should only exist to show demo content. These mockups violate that principle.

(BTW: This is an accessibility issue that also affects everyone. This is what people (from all backgrounds and ability levels) often think typing: "What was I supposed to type here?!?")

So, yeah: This makes it more difficult. Sorry. :cry:

**Comment by andreasn** [comment]
>  Placeholders should not be used as labels and should only exist to show demo content.

Ah, I see. Makes sense. Back to the drawing board!

**Comment by andreasn** [comment]
Now with proper labels:
mockup

Also removed the label mask from IPv6, since it's all prefixes with IPv6 https://www.quora.com/Does-IPv6-have-a-subnet-mask

**Comment by garrett** [comment]
Oh, labels on top? I'll have to adjust `ct-form` to support this. :wink: (It's something I should've done already and have been meaning to do anyway. It's also something that already works in mobile mode, so I just have to adapt things a bit, possibly. Maybe.)

**Comment by andreasn** [comment]
> Oh, labels on top?

Yeah, I'm a bit on the fence about that, but I couldn't figure out any other good way to maintain the hierarchy. My other take was this:

Screenshot from 2019-07-11 12-46-05

**Comment by garrett** [comment]
Yeah, labels on top make sense.

I'll have to add support for it before this PR lands. This means next week at _earliest_. (Sorry.)

**Comment by skobyda** [comment]
I tried to quickly rework it to correspond to mockup, tell me if something is wrong.
Screenshot from 2019-07-12 10-19-34

**Comment by andreasn** [comment]
> I tried to quickly rework it to correspond to mockup, tell me if something is wrong.

That's awesome, thanks!

**Comment by skobyda** [comment]
So on functional side, I would say this PR is finished. I just need to write tests and document all the helpers functions which I have for that.
Here's a few warnings, let me know if you agree with them:

**Should not be empty:**
Screenshot from 2019-07-16 16-31-26

**Invalid IP:**
Screenshot from 2019-07-16 16-33-45

**DHCP range not withing sub network:**
Screenshot from 2019-07-16 16-37-11

Also @garrett heres a dialog in default state. The default IP address is one of "special ips" set aside for private networks. All that user has to do here is just type in Name and he has a very basic virtual network.
Screenshot from 2019-07-16 16-30-03

**Comment by skobyda** [comment]
On functional side, I would say this PR is finished. I just need to write tests and document all the helpers functions which I have for that.
Here's a few warnings, let me know if you agree with them:

Should not be empty:
Screenshot from 2019-07-16 16-31-26



Invalid IP:
Screenshot from 2019-07-16 16-37-11


DHCP range not withing sub network:
Screenshot from 2019-07-16 16-30-03

Also @garrett heres a dialog in default state. The default IP address is one of "special ips" set aside for private networks. All that user has to do here is just type in Name and user has a very basic virtual network.
Screenshot from 2019-07-16 16-30-03

**Comment by garrett** [comment]
@skobyda: Judging just from the screenshots: I like this!

I'm trying your branch now, and I get this error:

Screenshot_2019-07-17 Virtual Machines - Sunny

Also: Would there be some way where we can auto-generate a useable name for the network (as a suggestion)?

**Comment by skobyda** [comment]
> I'm trying your branch now, and I get this error:
Fixed it. 

> Also: Would there be some way where we can auto-generate a useable name for the network (as a suggestion)?
I cannot think of good way. Just as with virtual machines, where Name is arbitrary. It just needs to be unique so you have a way how to reference specific network.
I can think of only doing network1, network2, network3...

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 1]
```diff
@@ -0,0 +1,608 @@
...
+                </Select.SelectEntry>
+                <Select.SelectEntry data='IPv4 and IPv6' key='IPv4 and IPv6'>
+                    {_("IPv4 and IPv6")}
+                </Select.SelectEntry>
+            </Select.Select>
+        </React.Fragment>
+    );
+};
+
+const DhcpRow = ({ ipVersion, rangeStart, rangeEnd, expanded, onValueChanged, validationFailed }) => {
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Here you allow the dhcpv4 settings.
However libvirt allows you to configure dhcpv6 as well.
Virt-manager has clear the difference between the two.
Do you think this deserves some clarification?

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 2]
```diff
@@ -0,0 +1,608 @@
...
+                               value={rangeStart}
+                               onChange={e => onValueChanged('ipv' + ipVersion + 'DhcpRangeStart', e.target.value)}
+                               className='form-control' />
+                            { validationStart == 'error' &&
+                            <HelpBlock>
+                                <p className='text-danger'>{validationFailed['ipv' + ipVersion + 'DhcpRangeStart']}</p>
+                            </HelpBlock> }
+                        </FormGroup>
+                    </div>
+                    <div className='ct-form-layout'>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: I think we renamed the ct-form-layout to ct-form already.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 3]
```diff
@@ -0,0 +1,608 @@
...
+                <Modal.Body>
+                    {body}
+                </Modal.Body>
+                <Modal.Footer>
+                    {this.state.dialogError && <ModalError dialogError={this.state.dialogError} dialogErrorDetail={this.state.dialogErrorDetail} />}
+                    {this.state.createInProgress && <div className="spinner spinner-sm pull-left" />}
+                    <Button bsStyle='default' className='btn-cancel' onClick={ this.props.close }>
+                        {_("Cancel")}
+                    </Button>
+                    <Button bsStyle='primary' disabled={ this.state.createInProgress } onClick={ this.onCreate }>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Please disable the button also if validation failed.

File: pkg/machines/libvirt-dbus.js [hunk 4]
```diff
@@ -1290,6 +1300,21 @@ export function networkUndefine(connectionName, objPath) {
     return call(connectionName, objPath, 'org.libvirt.Network', 'Undefine', [], TIMEOUT);
 }
 
+/* export function listInactiveInterfaces(connectionName) {
```
Comment (KKoukiou) [inline pkg/machines/libvirt-dbus.js]: You forgot this here I guess.

**Comment by KKoukiou** [comment]
@skobyda also the implementation is missing some network forward modes. For example routed. Is it planned to do it in separate PR?

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 5]
```diff
@@ -0,0 +1,608 @@
...
+    return (
+        <React.Fragment>
+            <label className='control-label'>
+                {_("Device")}
+            </label>
+            <FormGroup validationState={validationState} controlId='physicalDevice'>
+                <Select.Select id='create-network-physical-device'
+                               enabled={dialogValues.physicalDevice !== undefined}
+                               initial={dialogValues.physicalDevice !== undefined || "No Physical Device Available"}
+                               onChange={value => onValueChanged('physicalDevice', value)}>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: This is wrong. I guess you wanted to use short if then else. Aka condition ? value : other-value.

Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: Also this was breaking the select and you where not able to change from one device to another ;)

**Comment by KKoukiou** [comment]
> > I'm trying your branch now, and I get this error:
> 
> Fixed it.
> 
> > Also: Would there be some way where we can auto-generate a useable name for the network (as a suggestion)?
> 
> I cannot think of good way. Just as with virtual machines, where Name is arbitrary. It just needs to be unique so you have a way how to reference specific network.
> I can think of only doing network1, network2, network3... It wouldn't be hard to implement so I can implement it if you want.

I like the idea of autogenerating naming to prefill the dialog. (Not only this, in general of course) But I don't see some good enough naming approach that we can follow. When I create virtual networks, I usually name them according to their type. Like virt-net-routed-0 or virt-net-isolated so that I can remember how are they configured. 
But they type is configurable from the dialog, which makes the my naming convention not static. So I am not sure I would like some generic name refilled, like nework-1,2,3

File: pkg/machines/components/networks/createNetworkDialog.jsx (hunk 1 above)
Comment (skobyda) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: I don't know what clarification do you mean. This component is called from 2 places (IPv4Row and IPv6Row). I just use same component for that to not double code. 
Or do you mean make some clarification in the UI? In that case, I could change the text of related checkbox "Set DHCP Range" -> "Set IPvX DHCP Range".

**Comment by skobyda** [comment]
> @skobyda also the implementation is missing some network forward modes. For example routed. Is it planned to do it in separate PR?

Yes, I plan it for another PR, since this is already big enough. I would like to implement Bridge and VEPA modes next.

> I like the idea of autogenerating naming to prefill the dialog. (Not only this, in general of course) But I don't see some good enough naming approach that we can follow. When I create virtual networks, I usually name them according to their type. Like virt-net-routed-0 or virt-net-isolated so that I can remember how are they configured.
> But they type is configurable from the dialog, which makes the my naming convention not static. So I am not sure I would like some generic name refilled, like nework-1,2,3

Does it needs to be static? As long as user doesn't change name value for his own option, we could dynamically change name according to selected forward mode.

**Comment by skobyda** [comment]
I Added option of "Any physical device", which is basically not choosing any physical device and letting libvirt handle it.
In UI I used similar approach which we use in Create VM dialog for choosing a pool as destination storage. @garrett what do you think?
Screenshot from 2019-07-22 18-17-55

**Comment by marusak** [comment]
Needs to rebase to master since #12367 changed tests names

File: pkg/machines/components/networks/utils.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            return false;
+        const part = parseInt(ipv4[i], 10);
+        if (isNaN(part) || part < 0 || part > 255)
+            return false;
+    }
+
+    return true;
+}
+
+const netmaskParts = ["255", "254", "252", "248", "240", "224", "192", "128", "0"];
```
Comment (KKoukiou) [inline pkg/machines/components/networks/utils.js]: I don't see the reason for it to be outside of the function body

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 7]
```diff
@@ -0,0 +1,560 @@
...
+        if (dialogValues.ipv4DhcpEnabled) {
+            if (isEmpty(dialogValues.ipv4DhcpRangeStart.trim())) {
+                validationFailed['ipv4DhcpRangeStart'] = _("Start should not be empty");
+            } else if (!utils.validateIpv4(dialogValues.ipv4DhcpRangeStart)) {
+                validationFailed['ipv4DhcpRangeStart'] = _("Invalid IPv4 address");
+            } else if (!utils.isIpv4InNetwork(dialogValues.ipv4, dialogValues.netmask, dialogValues.ipv4DhcpRangeStart)) {
+                validationFailed['ipv4DhcpRangeStart'] = _("Address not within subnet");
+            }
+
+            if (isEmpty(dialogValues.ipv4DhcpRangeEnd.trim())) {
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: In general we don't use curly brackets for one liners.
If even one body of the condition clause is though multiline we use brackets.
I don't really care tbh, so you can leave it as is.
It's just a common convention.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+        </React.Fragment>
+    );
+};
+
+const NetworkNameRow = ({ onValueChanged, dialogValues, validationFailed }) => {
+    const validationState = validationFailed.name ? 'error' : undefined;
+
+    return (
+        <React.Fragment>
+            <label className='control-label'>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: You forgot htmlFor

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+            </FormGroup>
+        </React.Fragment>
+    );
+};
+
+const NetworkForwardModeRow = ({ onValueChanged, dialogValues }) => {
+    const forwardModes = ['nat', 'open', 'none'];
+
+    return (
+        <React.Fragment>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Missing htmlFor

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 10]
```diff
@@ -0,0 +1,560 @@
...
+        </React.Fragment>
+    );
+};
+
+const NetworkDeviceRow = ({ devices, onValueChanged, dialogValues, validationFailed }) => {
+    const validationState = validationFailed.physicalDevice ? 'error' : undefined;
+
+    return (
+        <React.Fragment>
+            <label className='control-label'>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Missing htmlFor

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 11]
```diff
@@ -0,0 +1,560 @@
...
+                </HelpBlock> }
+            </FormGroup>
+        </React.Fragment>
+    );
+};
+
+const IpRow = ({ onValueChanged, dialogValues }) => {
+    return (
+        <React.Fragment>
+            <label className='control-label'>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Same, htmlFor

File: pkg/machines/components/networks/createNetworkDialog.jsx (hunk 1 above)
Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: Sorry, didn't notice you reuse it.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 12]
```diff
@@ -0,0 +1,560 @@
...
+                    type='checkbox'
+                    checked={expanded}
+                    onChange={e => onValueChanged('ipv' + ipVersion + 'DhcpEnabled', !expanded)} />
+                {_("Set DHCP Range")}
+            </label>
+
+            {expanded && <React.Fragment>
+                <div className='create-network-dialog-grid'>
+                    <div className='ct-form'>
+                        <label className='control-label'> {_("Start")} </label>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: htmlFor. I will not mention this comment again. Please adjust similarly or missing htmlFor attrs.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 13]
```diff
@@ -0,0 +1,560 @@
...
+};
+
+const Ipv4Row = ({ validationFailed, dialogValues, onValueChanged }) => {
+    const validationAddress = validationFailed.ipv4 ? 'error' : undefined;
+    const validationNetmask = validationFailed.netmask ? 'error' : undefined;
+
+    return (
+        <React.Fragment>
+            <div className='create-network-dialog-grid'>
+                <div>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Is this div really needed. Maybe you can replace it with <React.Fragment> ?
@garrett really hates extra unneeded divs :)

Comment (skobyda) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: Oh this is just temporary until ct-form supports labels on top. It was mentioned in @garrett  comment when talking about design with andreas.

**Comment by garrett** [comment]
> I Added option of "Any physical device", which is basically not choosing any physical device and letting libvirt handle it.

"Any physical device" sounds like it would either pick a device at random, at some interval, or by some other means. It sounds magical (but not in a good way here).

What exactly does libvirt do when the physical device is not specified?

**Comment by skobyda** [comment]
> "Any physical device" sounds like it would either pick a device at random, at some interval, or by some other means. It sounds magical (but not in a good way here).

With "Any physical deice" option, routing table is used to dynamically select device which will be used. It uses various metrics to determine specific device to choose.(https://en.wikipedia.org/wiki/Metrics_(networking)).

**Comment by garrett** [comment]
I still don't understand fully. Does it:
- check the system and select a device up-front while the dialog is visible
- pick a different device from time to time
- use whatever

Is it like when I have both a wifi connection, a cat-5 cable plugged into my laptop, and VPN running and then I tell my laptop to connect to GitHub... and it just does it somehow and I don't care how? (And then I tell my computer to connect to my NAS and it uses the cable. And I care then.)

Is it the "System routing" or "Any device"? Perhaps even something like "Any device (default system routing)" — although this might start to become a little too wordy.

**Comment by skobyda** [comment]
> I still don't understand fully. Does it:
> 
> * check the system and select a device up-front while the dialog is visible
> * pick a different device from time to time
> * use whatever

It picks different device from time to time. Everytime packet is sent, kernel decides which device will be chosen.

> Is it like when I have both a wifi connection, a cat-5 cable plugged into my laptop, and VPN running and then I tell my laptop to connect to GitHub... and it just does it somehow and I don't care how? (And then I tell my computer to connect to my NAS and it uses the cable. And I care then.)
> 
> Is it the "System routing" or "Any device"? Perhaps even something like "Any device (default system routing)" — although this might start to become a little too wordy.

If you have multiple connections thru multiple devices as you described, than it works similar as when any process, any application in your laptop tries to connect to internet.

If some application on your laptop tries to send packet, kernel decides which network device shall be used by looking into routing table. There each interface has "Metric" value, and interface is chosen according to this value (highest one wins).

Virtual NAT network is same. If anything from this virtual network tries to send a packet, then kernel (after doing many other things) has to decide which interface is chosen. And it decides again by using routing table, in same manner as described in previous paragraph. This is the option "Any physical device".

Then you also have an option to specify network device for this virtual NAT network. In that case, libvirt will tell kernel that all packets from this virtual network shall go only thru this specific device. This is the option where you choose some specific physical device.

**Comment by garrett** [comment]
Would "Automatic" for the device option make sense?

Is that understandable to our users* and would it describe the situation well enough?

_* it's a good thing to ask in a usability study — until then, we're just guessing_

**Comment by skobyda** [comment]
> Would "Automatic" for the device option make sense?
> 
> Is that understandable to our users* and would it describe the situation well enough?
> 
> _* it's a good thing to ask in a usability study — until then, we're just guessing_

Automatic describes situation well, since kernel chooses that network device "automatically" in this case without any user input. 
I can imagine that option "automatic" next to list of devices which can user choose from could be interpreted as "device from this list will be chosen automatically", which is what actually happens.
And for user who has no knowledge about this topic, "Automatic" sound as safe, defaultish option, no? :D
In short, "Automatic" sounds good :).

**Comment by garrett** [comment]
> option "automatic" next to list of devices

It _also_ says "Device" to the left of the select option "Automatic", so it strongly implies an automatic device selection. :grin: But, yes, being an option in a list of devices also reinforces it by context.

So "Automatic" it is! :+1: 

(...At least unless a usability study indicates it needs changing. :wink:)

**Comment by skobyda** [comment]
So I added "Automatic" as our world.
Screenshot from 2019-08-01 14-54-37
Codewise, nothing has changed since @KKoukiou last review, except for our network devices getter . For that I will get libvirt guy to review it as @KKoukiou requested.

**Comment by KKoukiou** [comment]
> This is exactly how virt-manager creates the list of interfaces. Looking good.

Thanks Michal!

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+                    );
+                })
+                }
+            </Select.Select>
+        </React.Fragment>
+    );
+};
+
+const NetworkDeviceRow = ({ devices, onValueChanged, dialogValues, validationFailed }) => {
+    const validationState = validationFailed.device ? 'error' : undefined;
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: This will never be true, because you never set validationFailed for devices.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+                        { devices.map(dev => {
+                            return (
+                                <Select.SelectEntry data={dev} key={dev}>
+                                    {dev}
+                                </Select.SelectEntry>
+                            );
+                        })}
+                    </optgroup>
+                </Select.Select>
+                { validationState == 'error' &&
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: Because of the above comment this can be emitted.

File: pkg/machines/components/networks/createNetworkDialog.css [hunk 16]
```diff
@@ -0,0 +1,11 @@
+.create-network-dialog-grid {
+  display: grid;
+  grid-template-columns: 1fr 1fr;
+  grid-row-gap: 1em;
+  grid-column-gap: 1em;
```
Comment (garrett) [inline pkg/machines/components/networks/createNetworkDialog.css]: These can be consolidated to `grid-gap: 1em 1em;`, which can further be simplifed as `grid-gap: 1em;`

File: pkg/machines/components/networks/createNetworkDialog.css [hunk 17]
```diff
@@ -0,0 +1,11 @@
...
+  align: left;
```
Comment (garrett) [inline pkg/machines/components/networks/createNetworkDialog.css]: What's this?

Is it supposed to be `text-align: left;`, which is probably already the default? Or something else?

(As it isn't doing anything, you can remove this line.)

File: pkg/machines/components/networks/createNetworkDialog.jsx (hunk 14 above)
Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: Pushed the fix.

File: pkg/machines/components/networks/createNetworkDialog.jsx (hunk 15 above)
Comment (KKoukiou) [reply pkg/machines/components/networks/createNetworkDialog.jsx]: Pushed the fix.

File: pkg/machines/components/networks/createNetworkDialog.css [hunk 18]
```diff
@@ -0,0 +1,11 @@
+.create-network-dialog-grid {
+  display: grid;
+  grid-template-columns: 1fr 1fr;
```
Comment (garrett) [inline pkg/machines/components/networks/createNetworkDialog.css]: Also, you may want to consider adding `ct-form` to `create-network-dialog-grid` or changing the `1fr 1fr` to `auto 1fr` here.

Doing either should make it so there isn't such a big gap between the labels and the forms.

**Comment by skobyda** [comment]
To unblock this PR, we have decided to not use labels on top. Also resolved @garrett comments on css.
Current state of PR:
Screenshot from 2019-08-06 12-19-13

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 19]
```diff
@@ -0,0 +1,539 @@
...
+                        </Select.SelectEntry>
+                    );
+                })
+                }
+            </Select.Select>
+        </React.Fragment>
+    );
+};
+
+const NetworkDeviceRow = ({ devices, onValueChanged, dialogValues, validationFailed }) => {
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: validationFailed is not used. I don't know why React doesn't complain !? but it should be removed from the properties.

File: pkg/machines/components/networks/createNetworkDialog.jsx [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+                             validationFailed={validationFailed} /> }
+                </React.Fragment> }
+            </form>
+        );
+
+        return (
+            <Modal id='create-network-dialog' className='network-create' show onHide={ this.props.close }>
+                <Modal.Header>
+                    <Modal.CloseButton onClick={ this.props.close } />
+                    <Modal.Title> {`Create Virtual Network`} </Modal.Title>
```
Comment (KKoukiou) [inline pkg/machines/components/networks/createNetworkDialog.jsx]: This needs to be translated.

File: pkg/machines/components/networks/utils.js [hunk 21]
```diff
@@ -86,11 +85,11 @@ export function netmaskConvert(prefixOrNetmask) {
     const parts = prefixOrNetmask.split('.');
 
     if (parts.length === 4)
-        return netmask;
+        return prefixOrNetmask;
 
     const prefixLength = parseInt(parts[0]);
 
-    netmask = "";
+    let netmask = "";
```
Comment (KKoukiou) [inline pkg/machines/components/networks/utils.js]: I think these are reminders that should go in the first commit of this PR?

**Comment by skobyda** [comment]
> Two minor comments, otherwise looks good, thanks :)

Fixed.

**Comment by skobyda** [comment]
No machines related failures.

**Comment by KKoukiou** [comment]
Screen Shot 2019-08-07 at 17 32 45
